src/partial_quicksync.rs

```diff
@@ -0,0 +1,265 @@
+use anyhow::{Context, Result};
+// use parse_display::{Display, FromStr};
+use reqwest::blocking::Client;
+use rusqlite::Connection;
+use std::{
+  env,
+  fs::{self, File},
+  io::{self, BufReader, BufWriter},
+  str::FromStr,
+  time::Instant,
+};
+use zstd::stream::Decoder;
+
+const DEFAULT_BASE_URL: &str = "https://quicksync.spacemesh.network/partials";
+
+#[derive(Clone, Debug, parse_display::FromStr)]
+#[display("{from},{to},{hash}")]
+struct RestorePoint {
+  from: i64,
+  to: i64,
+  hash: String,
+}
+
+fn get_base_url() -> String {
+  env::var("QS_BASE_URL").unwrap_or_else(|_| DEFAULT_BASE_URL.to_string())
+}
+
+fn get_previous_hash(layer_at: i64, conn: &Connection) -> Result<String> {
+  conn
+    .query_row(
+      "SELECT aggregated_hash FROM layers WHERE id = ?",
+      [layer_at - 1],
+      |row| {
+        let hash: Vec<u8> = row.get(0)?;
+        Ok(hex::encode(&hash[..2]))
+      },
+    )
+    .context("Failed to get previous hash")
+}
+
+fn find_start_points(layer_from: i64, metadata: &str, jump_back: usize) -> Vec<RestorePoint> {
+  let mut result = Vec::new();
+
+  let mut target_index = 0;
+
+  for (index, line) in metadata.lines().enumerate() {
+    let items: Vec<&str> = line.split(',').collect();
+    let to: i64 = items[1].parse().unwrap();
+    if to > layer_from {
+      target_index = index;
+      break;
+    }
+  }
+  target_index = target_index - jump_back;
```

Comment:
It will panic if `jump_back > target_index` like in this test case:

```rust
let metadata = "0,100,aaaa\n101,200,bbbb\n201,300,ijkl\n";
let result = find_start_points(50, metadata, 1);
```